### PR TITLE
Add `odeint` kwargs in `ParametrizedEvolution`

### PR DESCRIPTION
**Pull request by AlbertMitjans**

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-32507]

File: tests/ops/op_math/test_parametrized_evolution.py [hunk 1]
```diff
@@ -95,6 +95,15 @@ def test_init(self):
...
         assert ev.num_params == 0
 
+    def test_odeint_kwargs(self):
+        """Test the initialization with odeint kwargs."""
+        ops = [qml.PauliX(0), qml.PauliY(1)]
+        coeffs = [1, 2]
+        H = ParametrizedHamiltonian(coeffs, ops)
+        ev = ParametrizedEvolution(H=H, params=[1, 2], t=2, mxstep=10)
+
+        assert ev.odeint_kwargs == {"mxstep": 10}
```
Comment (Qottmann) [inline tests/ops/op_math/test_parametrized_evolution.py:105]: Please do that for all kwargs

File: pennylane/ops/op_math/parametrized_evolution.py [hunk 2]
```diff
@@ -96,6 +96,13 @@ class ParametrizedEvolution(Operation):
         do_queue (bool): determines if the scalar product operator will be queued. Default is True.
         id (str or None): id for the scalar product operator. Default is None.
 
+    Kwargs:
```
Comment (Qottmann) [inline pennylane/ops/op_math/parametrized_evolution.py:99]: Is it typical to set a separate category for kwargs? Havent seen that in the codebase before but I like the idea

**Review by Qottmann — COMMENTED** [review]
Thanks @AlbertMitjans 👍 

Can we also add this functionality for `qml.evolve` or are there any complications?

File: tests/ops/op_math/test_parametrized_evolution.py (hunk 1 above)
Comment (AlbertMitjans) [reply tests/ops/op_math/test_parametrized_evolution.py:105]: Done!

File: pennylane/ops/op_math/parametrized_evolution.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/ops/op_math/parametrized_evolution.py:99]: I've seen it in some operator arithmetic class and I liked the idea

**Comment by AlbertMitjans** [comment]
> Can we also add this functionality for qml.evolve or are there any complications?

I'd love to discuss this more. Given that we want `qml.evolve` to be generic (such that it might return an `Evolution` or a `ParametrizedEvolution` operator) I don't like putting kwargs in this function that are not generic (such as the `odeint` kwargs). 

However, the user can still do:
```pycon
>>> qml.evolve(H)(params, t, atol=1e-6)
```

**Comment by lillian542** [comment]
`>>> qml.evolve(H)(params, t, atol=1e-6)`

Would be nice to include an example of this both here and in `qml.evolve`

File: pennylane/ops/op_math/parametrized_evolution.py [hunk 3]
```diff
@@ -96,6 +96,12 @@ class ParametrizedEvolution(Operation):
         do_queue (bool): determines if the scalar product operator will be queued. Default is True.
         id (str or None): id for the scalar product operator. Default is None.
 
+    Keyword Args:
+        rtol (float, optional): relative local error tolerance for solver. Defaults to 1.4e-8.
+        atol (float, optional): absolute local error tolerance for solver. Defaults to 1.4e-8.
```
Comment (Qottmann) [inline pennylane/ops/op_math/parametrized_evolution.py:101]: I would add how the error is computed and what the absolute and relative tolerance mean. Something like the following where they are put together (saw this in the scipy docs):
"rtol, atol (float): Relative and absolute error tolerance. The error is estimated from comparing a 4th and 5th order Runge-Kutta step in the Dopri5 algorithm. This error is guaranteed to stay below ``tol = atol + rtol * abs(y)`` through adaptive step size selection."

**Comment by Qottmann** [comment]
I see your point, maybe @trbromley has a suggestion what best to do. Since we are advertising the user to use ``evolve`` rather than `ParametrizedEvolution` I fear not giving the option in ``evolve`` would kind of bury this option.

> because then, in the following situation the kwargs would do nothing:
As long as it is just doing nothing and not breaking I would actually be fine with that.

**Comment by AlbertMitjans** [comment]
> Would be nice to include an example of this both here and in qml.evolve

Added examples!

File: pennylane/ops/op_math/parametrized_evolution.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/ops/op_math/parametrized_evolution.py:114]: Added!

**Comment by AlbertMitjans** [comment]
> I fear not giving the option in evolve would kind of bury this option

Why would it bury this option? Adding the kwargs in `evolve` would mean that the user can do:

```pycon
>>> qml.evolve(H, atol=1e-4)(params, t)
```

Instead of

```pycon
>>> qml.evolve(H)(params, t, atol=1e-4)
```

For just this little change, I'd prefer keeping the kwargs inside `ParametrizedEvolution` to avoid any confusion

File: pennylane/ops/op_math/parametrized_evolution.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/ops/op_math/parametrized_evolution.py]: Just a minor thing, it is not rendering as expected: 
image

**Review by Qottmann — APPROVED** [review]
Ah I misunderstood, I thought a user would have to use `ParametrizedEvolution(H, params, t, atol=1e-5)` in order to access the kwargs functionality and it would be impossible to do on the `qml.evolve` level. All good then! 👍

File: pennylane/ops/op_math/parametrized_evolution.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/ops/op_math/parametrized_evolution.py]: Fixed!

File: pennylane/ops/op_math/parametrized_evolution.py [hunk 4]
```diff
@@ -84,18 +89,36 @@ class ParametrizedEvolution(Operation):
...
         time (str, optional): The name of the time-based parameter in the parametrized Hamiltonian.
             Defaults to "t".
         do_queue (bool): determines if the scalar product operator will be queued. Default is True.
         id (str or None): id for the scalar product operator. Default is None.
 
+    Keyword Args:
+        atol (float, optional): Absolute error tolerance. Defaults to 1.4e-8.
+        rtol (float, optional): Relative error tolerance. The error is estimated
+            from comparing a 4th and 5th order Runge-Kutta step in the Dopri5 algorithm. This error
+            is guaranteed to stay below tol = atol + rtol * abs(y) through adaptive step size
```
Comment (Qottmann) [inline pennylane/ops/op_math/parametrized_evolution.py:116]:
```suggestion
            is guaranteed to stay below ``tol = atol + rtol * abs(y)`` through adaptive step size
```

**Pull request merged**